src/main/scala/org/apache/spark/sql/delta/storage/GcsLogStore.scala

```diff
@@ -0,0 +1,97 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.storage
+
+import java.io.{IOException, _}
+import java.nio.charset.StandardCharsets.UTF_8
+import java.nio.file.FileAlreadyExistsException
+
+import com.google.common.base.Throwables
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+
+/**
+ * The [[LogStore]] implementation for GCS, which uses gcs-connector to
+ * provide the necessary atomic and durability guarantees:
+ *
+ * 1. Atomic Visibility: Read/read-after-metadata-update/delete are strongly
+ * consistent for GCS.
+ *
+ * 2. Consistent Listing: GCS guarantees strong consistency for both object and
+ * bucket listing operations.
+ * https://cloud.google.com/storage/docs/consistency
+ *
+ * 3. Mutual Exclusion: Preconditions are used to handle race conditions.
+ *
+ * Regarding file creation, this implementation:
+ * - Opens a stream to write to GCS otherwise.
+ * - Throws [[FileAlreadyExistsException]] if file exists and overwrite is false.
+ * - Assumes file writing to be all-or-nothing, irrespective of overwrite option.
+ */
+class GcsLogStore(sparkConf: SparkConf, defaultHadoopConf: Configuration)
+  extends HadoopFileSystemLogStore(sparkConf, defaultHadoopConf) with Logging {
+
+  val preconditionFailedExceptionMessage = "412 Precondition Failed"
+
+  def write(path: Path, actions: Iterator[String], overwrite: Boolean = false): Unit = {
+    val fs = path.getFileSystem(getHadoopConfiguration)
+
+    // This is needed for the tests to throw error with local file system.
+    if (fs.isInstanceOf[LocalFileSystem] && !overwrite && fs.exists(path)) {
+      throw new FileAlreadyExistsException(path.toString)
+    }
+
+    try {
+      // If overwrite=false and path already exists, gcs-connector will throw
+      // org.apache.hadoop.fs.FileAlreadyExistsException after fs.create is invoked.
+      // This should be mapped to java.nio.file.FileAlreadyExistsException.
+      val stream = fs.create(path, overwrite)
+      try {
+        actions.map(_ + "\n").map(_.getBytes(UTF_8)).foreach(stream.write)
+      } finally {
+        stream.close()
+      }
+    } catch {
+      case e: org.apache.hadoop.fs.FileAlreadyExistsException =>
+        throw new FileAlreadyExistsException(path.toString).initCause(e.getCause)
+      // GCS uses preconditions to handle race conditions for multiple writers.
+      // If path gets created between fs.create and stream.close by an external
+      // agent or race conditions. Then this block will execute.
+      // Reference: https://cloud.google.com/storage/docs/generations-preconditions
+      case e: IOException if isPreconditionFailure(e) =>
+        if (!overwrite) {
+          throw new FileAlreadyExistsException(path.toString).initCause(e.getCause)
+        }
```

Comment:
Is there a way to validate/confirm that this is correct logic from Delta Lake API/contract point of view?

Reply:
I am not sure if there is one, at least for `overwrite=true`. I looked at this [doc][1] for concurrency logic.

I tested above by manually trying to trigger race condition. For `overwrite=false` deltalake existing logic worked as described in the doc(retried log store creation) and upon read all data(both manually inserted and retried) was available. 

Testing for `overwrite=true`, transaction was lost(as in not available for subsequent reads). Deltalake actually uses `overwrite=true` for checkpoint creation, which happens after a transaction gets committed successfully. So I'm not sure if contention can happen in that case. 

Maybe we can write our own test case in examples/python if you want but then again it won't verify if we're actually following contract or not. Ongoing PRs on deltalake also don't offer anything new, [oracle-cloud][2] and [dynamodb][3]. I'll go through other unit tests to find something helpful.

[1]: https://databricks.com/blog/2019/08/21/diving-into-delta-lake-unpacking-the-transaction-log.html
[2]: https://github.com/delta-io/delta/pull/468
[3]: https://github.com/delta-io/delta/pull/339

Reply:
Thanks for the investigation. I think that the best option will be to make GCS implementation consistent with HDFS implementation.

May you check how concurrent overwrites behave on HDFS? And if they do not fail, I think we are good to go.

Reply:
Concurrent overwrites are not an issue with `HDFSLogStore`. Logical flow for it is: 

1) write data to temporary file
2) use `FileContext.rename` to rename temporary file to actual file

Now for `overwrite=false` if we create a file between 1 and 2. It throws `FileAlreadyExistsException` for append-type transaction and retries, whereas for overwrite-type transaction it fails with `ConcurrentAppendException`. This is consistent with `GcsLogStore`.

For `overwrite=true` if we create a file between 1 and 2 it simply overwrites it, as there is no proper way to simulate race conditions. Also `FileContext.rename` invokes `FileSystem.rename`, so even though it is atomic for HDFS, same won't hold true for Gcs. 

Regarding suppression of exception for `overwrite=true` in `GcsLogStore`, `HDFSLogStore` doesn't suppresses any exception which it is not expecting, i.e., `org.apache.hadoop.fs.FileAlreadyExistsException` [reference][1]. For example, if we manually delete temporary file between 1 and 2, it throws FileNotFoundException, so maybe we should also just IOException as it is for precondition-failure in case of `overwrite=true`.

[1]: https://github.com/delta-io/delta/blob/master/src/main/scala/org/apache/spark/sql/delta/storage/HDFSLogStore.scala#L98-L102

Reply:
I checked [`FileContext.rename` javadoc](https://hadoop.apache.org/docs/r3.3.0/api/org/apache/hadoop/fs/FileContext.html#rename-org.apache.hadoop.fs.Path-org.apache.hadoop.fs.Path-org.apache.hadoop.fs.Options.Rename...-) - in overwrite mode it does not fail, so current `GcsLogStore` implementation is consistent with this behavior.